### ISSUE TITLE
Intermediate Algorithm Search And Replace needs to have solution replaced

### ISSUE THREAD

**Issue by stevensacks** [freeCodeCamp/freeCodeCamp#34761]
https://guide.freecodecamp.org/certifications/javascript-algorithms-and-data-structures/intermediate-algorithm-scripting/search-and-replace/

`Advanced Code Solution Alternative` is not an acceptable solution for numerous reasons.
1. It adds a prototype function to a native Javascript object. This is a no-no. It should never be done.
2. It uses an IIFE for no reason, and there is nothing in the curriculum up to this point that explains what an IIFE is, why you would use an IIFE (and in this case, you wouldn't), and doesn't even label it as an IIFE, so it just looks like weird syntax to learners.
3. It does not follow KISS. It's overly complex in numerous places. It's inelegant.
4. It uses `else return` after the if returns, which is unnecessary.
5. It provides no code explanation below, which is especially bad because it uses an IIFE.

The bottom line is, it is not an advanced solution. This type of code is often written by intermediate programmers who think they're advanced programmers because they overcomplicate everything and think they're being clever. Sure, it's a phase we all go through, but an advanced developer would not write code like this.

I am submitting an advanced solution and this issue will be referenced by the PR.

**Comment by stevensacks** [comment]
Note that the "Run Code" link for my solution is not correct in the PR. I don't know how to set what that link is.

**Comment by ezioda004** [comment]
Hi @stevensacks,

Yes, ofcourse *overwriting* a method to the prototype is a bad practice. However, the solution is not overwriting. This is a common way of adding a method if it *doesnt* exists. Matter of fact thats basically how polyfill works.

The IIFE point you mentioned I understand where its coming from. However, the solution is going for the 
modular kinda pattern - which most popular libraries like jQuery/lodash use.

That being said - Advanced solutions IMO are the ones which are objectively better (in terms of space and time complexity). Previously discussed with @RandellDawson - an advanced can also be the solution which teaches new things/practices. 
A beginner might not be able to understand the solution at first glance, but thats the point. It gives you new insights and helps you think about new ways to approach a problem.


The other criticisms are valid. The code can use more explanation regarding the use of IIFE.

**Comment by stevensacks** [comment]
The only place it's common is junior devs who don't know better, hacks who don't care, and script kiddies who are just messing around. 

Polyfill is the **only** place that adding to a native object prototype is acceptable because all that a Polyfill does is add functions that are **expected** to be supported in the spec.

There's a reason why **every Javascript library** does not do it.

Lodash, Underscore, jQuery, etc. do not add functions to native object prototypes. There are a ton of downsides. I could get into them all but it's not worth it. Just don't do it. The only upside is selfish.

https://javascript.info/native-prototypes#native-prototype-change
>During the process of development we may have ideas which new built-in methods we’d like to have. And there may be a slight temptation to add them to native prototypes. But that is generally a bad idea.
>
>Prototypes are global, so it’s easy to get a conflict. If two libraries add a method String.prototype.show, then one of them overwrites the other one.
>
>In modern programming, there is only one case when modifying native prototypes is approved. >That’s polyfills. In other words, if there’s a method in JavaScript specification that is not yet >supported by our JavaScript engine (or any of those that we want to support), then may implement >it manually and populate the built-in prototype with it.

Install this rule and be done with it:
https://eslint.org/docs/rules/no-extend-native

And finally, I disagree with advanced solutions that "implicitly" teach new things/practices. For people who are learning, they're at best a distraction and considering that there seems to be little to no review for bad practices and/or sufficient documentation and/or links for more information to these new things/practices, they're confusing and harmful.

IIFEs are barely used anymore. And they're a fairly esoteric part of the Javascript spec. Somebody who's just learning how to do intermediate algorithms in Javascript has absolutely no use for IIFEs, and they don't have the relevant context for what IIFEs are and why you would use one.

If you're learning how to play piano and you've memorized a few chord progressions for some of your favorite pop songs and are learning about major and minor scales, and I throw in a lesson about the importance of Subdominants in Ionian mode, you wouldn't have any idea what I was talking about. Even if I explain it, there's a ton of context in between what you currently know and what I'm trying to teach you. It will only confuse you.

An advanced example should not teach esoteric things. It shouldn't use techniques that are frowned upon. It shouldn't be overly complicated and verbose. It should take the easy answer and show how it can be done more elegantly. If there's some new technique that is being used that is potentially unclear, then a link should be provided for those who want clarity. But those techniques should be applicable in real-life and in following tutorials.

My code is, I think, a great example of what an advanced solution should be. My friend is a super beginner and even he understood what I did. He was able to look at it and say "whoah! I get it. I hadn't thought that way but I could". Instead of "wtf is going on here? I don't understand this at all."

```javascript
const isLetter = char => char.match(/^[a-zA-Z]*$/);

const isUpperCase = char => char.charCodeAt(0) > 64;

// if A is not a letter, B case does not change
const matchCharCase = (a, b) => {
    if (isLetter(a)) {
        if (isUpperCase(a)) return b.toUpperCase();
        return b.toLowerCase();
    }
    return b;
};

// 1. Trim A to the length of B
// 2. Split into array and map over each character
// 3. Match case for each character
// 4. Join result back into a string
// 5. If B is longer than A, append remainder of B

const matchCase = (a, b) => a
    .slice(0, b.length) 
    .split('')
    .map((c, i) => matchCharCase(c, b[i]))
    .join('') + b.slice(a.length);

const myReplace = (str, searchFor, replaceWith) => 
    str.replace(searchFor, matchCase(searchFor, replaceWith));
```

**Comment by stevensacks** [comment]
I'm going to provide a breakdown piece by piece of the previous advanced example just like I would in a code review.

**ORIGINAL**
```javascript
// Add new method to the String object, not overriding it if one exists already
String.prototype.capitalize =  String.prototype.capitalize ||
    function() {
        return this[0].toUpperCase() + this.slice(1);
    };
```

It's confusing to mix ES6 terms (const) and ES5 terms (function). Because we are teaching ES6, and this test comes after the Functional Programming section, we should avoid using the `function` keyword or using `this`.

You shouldn't add to native prototypes. Write an arrow function instead.

**REFACTOR**
```javascript
const capitalize = str => str[0].toUpperCase() + str.slice(1);
```

**ORIGINAL**
```javascript
const Util = (function () {
// Create utility module to hold helper functions
    function textCase(str, tCase) {
        // Depending if the tCase argument is passed we either set the case of the
        // given string or we get it.
        // Those functions can be expanded for other text cases.
        
        if(tCase) {
            return setCase(str, tCase);
        } else {
            return getCase(str);
        }

        function setCase(str, tCase) {
            switch(tCase) {
                case "uppercase": return str.toUpperCase();
                case "lowercase": return str.toLowerCase();
                case "capitalized": return str.capitalize();
                default: return str;
            }
        }

        function getCase(str) {
            if (str === str.toUpperCase()) { return "uppercase"; }
            if (str === str.toLowerCase()) { return "lowercase"; }
            if (str === str.capitalize()) { return "capitalized"; }
            return "normal";
        }
    }

    return {
        textCase
    };
})();
```

The IIFE serves no purpose here other than to obfuscate the code. This Util function could be achieved with properties on an object, but is actually not necessary at all. It's preferable to use independent functions to avoid potential scope issues here or down the line. If you want to make a Utility module for Strings, export each const individually from the module or export a default object leveraging ES6 object destructuring.

The functions themselves are left out as this is only to show structure. There are much simpler ways to accomplish the things these functions do.

**REFACTOR**
```javascript
const setCase = (str, tCase) => // ...
const getCase = (str) => //...
const textCase = (str, tCase) => tCase ? setCase(str, tCase) : getCase(str);

export default {
    setCase,
    getCase,
    textCase,
};
```

**ORIGINAL**
```javascript
function myReplace(str, before, after) {
    const { textCase } = Util;
    const regex = new RegExp(before, 'gi');
    const replacingStr = textCase(after, textCase(before));

    return str.replace(regex, replacingStr);
}
```

The words `before` and `after` are unclear. Better to use descriptive argument names such as `searchFor` and `replaceWith`.

Put a comment to explain that the regex is for doing a case-insensitive search.

**REFACTOR**
```javascript
const myReplace = (str, searchFor, replaceWith) => {
   // do a case-insensitive search 
   const regex = new RegExp(before, 'gi');
   const matchedCase = textCase(after, textCase(before));
   return str.replace(regex, matchedCase);
};

// simple arrow function
const myReplace = (str, searchFor, replaceWith) => 
    str.replace(new RegExp(before, 'gi'), textCase(after, textCase(before)))
```

**Comment by ezioda004** [comment]
>In modern programming, there is only one case when modifying native prototypes is approved. That’s polyfills.

Fair point.

>IIFEs are barely used anymore. And they're a fairly esoteric part of the Javascript spec. Somebody who's just learning how to do intermediate algorithms in Javascript has absolutely no use for IIFEs, and they don't have the relevant context for what IIFEs are and why you would use one.

That is very opinionated. I'm sure many people who dont know a certain concept can/will google out of curiosity. I always liked the fact that FCC doesnt hold your hand to teach you things, it actually encourages you to do your own research.

FYI - FCC thrives on "Read, Search and Ask" methodology.

>...It should take the "Beginner" or "Intermediate" answer(s) and show how it can be done more elegantly and/or cleverly. 

Why wont you show the elegant/clever solution in the first go? Advanced solutions should be objectively better while using best practices (This one isnt).

>It's confusing to mix ES6 terms (const) and ES5 terms (function).

I've yet to meet a single programmer who have found using both of those terms together confusing.

>Because we are teaching ES6, and this test comes after the Functional Programming section, we should avoid using the function keyword or using this

This section also comes after basic JS/OOP section. Lets not favor one paradigm over the other. Its a multi-paradigm language. 

>The IIFE serves no purpose here other than to obfuscate the code. 

Its a design pattern called ["Revealing Module Pattern"](https://addyosmani.com/resources/essentialjsdesignpatterns/book/#revealingmodulepatternjavascript)


Before this goes to further bikeshed, I'll stop.

Currently, what goes into the guide or what is classified as "basic", "intermediate" or "advanced" solution is purely subjective. 

We're working on the guidelines to make it more explicit and objective.

**Comment by stevensacks** [comment]
The only reason I was able to identify that as an IIFE is because I know what an IIFE is. There’s no google translate that you can paste JS into that will tell you that (function(){})() is called an IIFE. They won’t be able to look it up if they don’t know what it’s called.

There’s nothing to tell a beginner what that pattern is called so there’s nothing for them to learn or grok from seeing it. They don’t know what they’re looking at, only that it’s confusing.

A beginner isn’t going to understand what Immediately Invoked Function Expression means, how it works, and most importantly, under which situations would you want to use it and why.

There is no benefit or value to using that pattern for this test and doing so is nothing but developer masturbation.

If you’re teaching ES6 then build ES6 habits. Use arrow functions, don’t use var, only use const (only use let if absolutely necessary). You’re not meeting a programmer in the wild. These are people new to programming. Build good habits from the start. Don’t lead them down paths they shouldn’t go down yet, if ever.

In my example, there is value in writing code declaratively, writing top to bottom and having each function do one thing (for the most part). Even if they don’t 100% understand how my final function works, they can follow the logic and learn to think like a programmer by breaking things into discrete pieces.

The fact that my friend could understand how my example worked right away speaks volumes. Even though my example looks simple, there’s quite a few advanced ways of thinking going on in there.

For example, a beginner might not know that if you slice a string to an index outside the length it returns just the string with no extra characters at the end. Or if you slice from an index outside the length it returns an empty string. But knowing that fact is different from seeing how they knowledge can be applied, even in this simple example.

I don’t think discussing effective (and ineffective) ways to teach students is bikeshedding.

**Comment by ezioda004** [comment]
>The only reason I was able to identify that as an IIFE is because I know what an IIFE is. There’s no google translate that you can paste JS into that will tell you that (function(){})() is called an IIFE. They won’t be able to look it up if they don’t know what it’s called.

We have gitter chat and forums for this reason only. If someone cant find something or is stuck - they can always ask others. Its a part of "Read, Search and **Ask**" :)

>There’s nothing to tell a beginner what that pattern is called so there’s nothing for them to learn or grok from seeing it.

As said previously, the example would benefit from an external link and comment about it.

>A beginner isn’t going to understand what Immediately Invoked Function Expression means, how it works, and most importantly, under which situations would you want to use it and why.

We have challenges on [IIFE](https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/object-oriented-programming/understand-the-immediately-invoked-function-expression-iife/) and [modular](https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/object-oriented-programming/use-an-iife-to-create-a-module/) pattern in previous sections. 

>If you’re teaching ES6 then build ES6 habits. Use arrow functions, don’t use var, only use const (only use let if absolutely necessary). You’re not meeting a programmer in the wild. These are people new to programming. Build good habits from the start. Don’t lead them down paths they shouldn’t go down yet, if ever.

Again, thats opinionated. The guide is made by thousands of contributors. We dont want to restrict anyone from using `function` just because arrow functions exist. Just because one have a preference of using "x" doesnt mean no one else can use "y".
As long as its not a bad practice. Its completely fine. 

>Little things like this expose valuable nuggets of knowledge about how JS works, instead of overwhelming grandiose things like design patterns and IIFEs. Just because you know your ABCs and what nouns, verbs, and prepositions are doesn’t mean you’re ready to read The Oddysey.

If they are in the last section of JS, using what has been taught previously as an example is fair.
One can always revisit the challenge to refresh their memory.
